cdap-app-fabric/src/main/java/io/cdap/cdap/app/job/SupportBundleJob.java

```diff
@@ -0,0 +1,471 @@
+import io.cdap.cdap.common.logging.LoggingContext;
+import io.cdap.cdap.internal.app.services.ApplicationLifecycleService;
+import io.cdap.cdap.internal.app.store.RunRecordDetail;
+import io.cdap.cdap.logging.context.LoggingContextHelper;
+import io.cdap.cdap.logging.read.LogReader;
+import io.cdap.cdap.metrics.query.MetricsQueryHelper;
+import io.cdap.cdap.proto.ApplicationDetail;
+import io.cdap.cdap.proto.ApplicationRecord;
+import io.cdap.cdap.proto.ProgramType;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.ProgramRunId;
+import javafx.util.Pair;
+import org.joda.time.DateTime;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+public class SupportBundleJob {
+  private static final Logger LOG = LoggerFactory.getLogger(SupportBundleJob.class);
+  private static final ExecutorService executor = Executors.newFixedThreadPool(5);
+  private static final SupportBundleSystemLogTask supportBundleSystemLogTask =
+      new SupportBundleSystemLogTask();
+  private static final SupportBundlePipelineInfoTask supportBundlePipelineInfoTask =
+      new SupportBundlePipelineInfoTask();
+  private static final SupportBundlePipelineRunLogTask supportBundlePipelineRunLogTask =
+      new SupportBundlePipelineRunLogTask();
+  private static final SupportBundleRuntimeInfoTask supportBundleRuntimeInfoTask =
+      new SupportBundleRuntimeInfoTask();
+  private static List<RunRecordDetail> runRecordDetailList;
+  private static Map<String, Integer> retryServiceMap;
+  private static Queue<String> queue;
+  private static Map<RunRecordDetail, JsonObject> runMetricsMap;
+  private static JsonArray tasks;
+  private final JsonObject statusJson;
+  private final Gson gson;
+  private final Store store;
+  private final ApplicationLifecycleService applicationLifecycleService;
+  private final MetricsQueryHelper metricsQueryHelper;
+  private final CConfiguration cConf;
+  private final LogReader logReader;
+
+  public SupportBundleJob(
+      CConfiguration cConf,
+      LogReader logReader,
+      List<String> serviceList,
+      ApplicationLifecycleService applicationLifecycleService,
+      Gson gson,
+      Store store,
+      MetricsQueryHelper metricsQueryHelper,
+      JsonObject statusJson) {
+    this.cConf = cConf;
+    this.logReader = logReader;
+    this.applicationLifecycleService = applicationLifecycleService;
+    this.gson = gson;
+    this.store = store;
+    this.metricsQueryHelper = metricsQueryHelper;
+    this.statusJson = statusJson;
+    retryServiceMap = new HashMap<>();
+    queue = new LinkedList<>();
```

Comment:
Don't use LinkedList as a queue. It's not thread safe. There are good thread-safe queue implementations